added init command

tdfranklin opened the pull request:
Ok, I wanted you to take a look at this before I went any further and get your opinions.

I've modified the CLI so that if you just type `reframe` or any argument except for `reframe start` or `reframe init` it will send back an error.

With `reframe start`, I simply changed the name of your run() function to start and moved it behind that command instead of running by default.

With `reframe init projectName` I have it creating the scaffold structure we discussed:

- projectName
  - app
    - views
      - homeView.js
    - pages
      - homePage.js

With that said, I think the createScaffold() function is a little messy with all the nested callbacks.  If you don't have a better suggestion, I think we should consider also including [fs-extra](https://www.npmjs.com/package/fs-extra) as a dependency since it is a direct replacement for Node's fs module and will recursively go back and create the file structure with a single command.  Just let me know if you want me to add that and I'll go back and clean that up.

Also, I don't have it creating a `package.json` file yet as I wanted to get you opinion on what you wanted it to look like.  We can discuss that later and figure out what all we want to include in that.

For the `homeView.js` and `homePage.js` I'm not sure if I did everything correctly to link those two or if I'm missing some other boilerplate code we want to add.  Just let me know what you think!

brillout commented on core/cli/templates/homeTemplate.js:
```diff
@@ -0,0 +1,47 @@
+`import React, { Component } from 'react';
+
+class HomeView extends Component {
+    constructor(props) {
```
I've checked what kind of components create-react-app scaffolds. It creates a ES6 class component without constructor and without defaultProps. I'd maybe do the same. In general I would tend to scaffold only a minimal amount of code. Feel free to contradict

tdfranklin replied:
I will certainly follow your lead on this.  In my limited experience I would think you would want the constructor function with state by default, but again, I haven't built a lot of React apps, so I will go ahead and drop it and defaultProps for now.  If at some point we decide we want to keep it (like if we end up wanting to load something in state for the entry point of the site) we can just add it back.

brillout commented on core/cli/cli.js:
```diff
@@ -118,3 +155,48 @@ function green_checkmark() {
+    const {homeViewTemplate, homePageTemplate} = require('./templates/homeTemplate');
+    const viewTemplate = homeViewTemplate();
+    const pageTemplate = homePageTemplate(projectName);
+    let currentDir = path.normalize(path.resolve(process.cwd(), projectName));
```
I guess we don't need `path.normalize` here. Correct me if I'm wrong.

tdfranklin replied:
You are probably correct.  I guess I was thinking that if someone wanted to pass an actual path as an argument, we may want to normalize it, but that's probably not something we need to support.  I'll remove it!

brillout commented on core/cli/cli.js:
```diff
@@ -118,3 +155,48 @@ function green_checkmark() {
+                        if (err) {
+                            console.log(err);
+                        } else {
+                            let tempPath = path.resolve(currentDir, 'app', 'pages', 'homePage.js');
```
It needs to be `homePage.html.js`. (I'm thinking of getting rid of the `.html.js` suffix requirement btw.)

tdfranklin replied:
Sorry about that!  I completely forgot about the .html part.  I'm adding for now and if you decide to remove that in the future, we can adjust.

brillout commented:
Looks really good, I'm excited.

> I think we should consider also including fs-extra

Sounds good, then we can do `await fs.ensureDir(directory)` :heart:

> Also, I don't have it creating a package.json file yet as I wanted to get you opinion on what you wanted it to look like. We can discuss that later and figure out what all we want to include in that.

I guess a minimal package.json would be
```json
{
  "name": "projectName",
  "dependencies": {
    "react": "^16.2.0"
  }
}
```

I like your PR, can't wait to review the following commits

brillout commented:
Also, I've created a ticket for this PR: https://github.com/brillout/reframe/issues/7

tdfranklin commented:
Awesome!  I'm going to add fs-extra and clean that function up and then add the package.json and resubmit!

tdfranklin commented:
Ok @brillout I've pushed another commit.

I added the `package.json` as you suggested and removed defaultprops and state from the viewTemplate.

Mainly, however, I cleaned up the `createScaffold()` function and I think it looks a lot better (but anxious to hear your thoughts).  I could easily squash that code into a single try/catch line and get rid of a lot of that code if you would prefer, but I thought maybe having a different try/catch block for each folder would make it easier to expand on later if we wanted to add additional files to any of those folders.  So right now, there's a block for `projectName` base folder, one for `projectName/app/views` and one for `projectName/app/pages` so we could easily add additional files if we needed to.

If you think that's overkill, though, just let me know and I'll be happy to squash it all into a single block.

brillout commented:
Awesome, only missing thing for this to work is https://github.com/brillout/reframe/issues/8 then it will work.

About the try/catch blocks: Actually, I don't think we need any of the try/catch blocks at all: If something fails the execution of the script stops, which is, I think, a good thing. With the try/catch blocks, if something fails, then the error will be `console.log` and the execution will resume.

Going to add 2-3 comments about some details about the code.

Almost there 😍

brillout commented on core/cli/cli.js:
```diff
@@ -1,10 +1,47 @@
+}
 
-run();
+//process.on('unhandledRejection', err => {throw err});
```
why comment this away?

tdfranklin replied:
Oh yea!  I wanted to ask you about that.  Still kind of learning about the way processes work and wasn't sure if we needed this with the error handling I passed through commander, but actually with the statements above, I think I see what's going on.  So that's basically just an event listener and if there's an error, it will stop the process and throw an error, correct?

Just out of curiosity, instead of throwing the error, should we just console.log it to make it look cleaner?  we could do something like:

```javascript
process.on('unhandledRejection', err => {
  console.log(err);
  process.exit(1);
}
```

If you prefer to throw, though, I'll follow your lead :)

brillout replied:
> So that's basically just an event listener and if there's an error, it will stop the process and throw an error, correct?

Yes. More precisely it is rejected promises that are listened for. But I guess that's what you were thinking.

And yes we could `console.log(err); process.exit(1);` instead. I just find that `throw err` expresses our intention slightly better. But that's nitpicking and either way is fine to me.

brillout commented on core/cli/templates/homeTemplate.js:
```diff
@@ -0,0 +1,39 @@
+    }
+}
+
+export default HomeView;`;
```
Files typically end with a new line

tdfranklin replied:
Oh really?!  I did not know that.  My OCD bugs me about empty lines at the end of a file, but I'll get those updated ASAP!

brillout replied:
You can `~ cat path/to/file/without/new/line` and see that the terminal won't start on a new line after the file is printed.

tdfranklin commented:
Ok, I think I made all changes.  See if that looks better now?
